### PR TITLE
System metrics

### PR DESCRIPTION
**Pull request by wubin01**
**What does this PR do?**
add system metrics

**Why are these changes required?**

**This PR has been tested by:**
- Unit Tests
- Manual Testing

**Follow up**

**Extra details**

### CODE REVIEW

**Comment by guoquanwu** [comment]
please check the code style

File: chainbase/src/main/java/org/tron/core/store/DynamicPropertiesStore.java [hunk 1]
```diff
@@ -1810,6 +1811,20 @@ public long getChangeDelegation() {
         .orElseThrow(() -> new IllegalArgumentException("not found CHANGE_DELEGATION"));
   }
 
+  /**
+   * block total processing transaction time
+   */
+  public void saveTotalProcessingTxTime(String bigInteger) {
```
Comment (spidemen) [inline chainbase/src/main/java/org/tron/core/store/DynamicPropertiesStore.java]: remove

Comment (Starsakary) [reply chainbase/src/main/java/org/tron/core/store/DynamicPropertiesStore.java]: ok

Comment (wubin01) [reply chainbase/src/main/java/org/tron/core/store/DynamicPropertiesStore.java]: ok

File: chainbase/src/main/java/org/tron/core/store/DynamicPropertiesStore.java [hunk 2]
```diff
@@ -1810,6 +1811,20 @@ public long getChangeDelegation() {
...
+    this.put(TOTAL_PROCESSINGTx_TIME, new BytesCapsule(ByteArray.fromHexString(bigInteger)));
+  }
+
+  public String getTotalProcessingTxTime() {
```
Comment (spidemen) [inline chainbase/src/main/java/org/tron/core/store/DynamicPropertiesStore.java]: remove this function

Comment (spidemen) [reply chainbase/src/main/java/org/tron/core/store/DynamicPropertiesStore.java]: remove

File: common/src/main/java/org/tron/common/entity/NodeInfo.java [hunk 3]
```diff
@@ -29,6 +29,8 @@
 
   private Map<String, String> cheatWitnessInfoMap = new HashMap<>();
 
+  private String totalProcessTxTime;
```
Comment (wubin01) [inline common/src/main/java/org/tron/common/entity/NodeInfo.java]: This field is not required

File: chainbase/src/main/java/org/tron/core/store/DynamicPropertiesStore.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (wubin01) [reply chainbase/src/main/java/org/tron/core/store/DynamicPropertiesStore.java]: No need to save to DB

File: common/src/main/java/org/tron/common/entity/NodeInfo.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (wubin01) [reply common/src/main/java/org/tron/common/entity/NodeInfo.java]: This field is not required

File: framework/src/main/java/org/tron/core/consensus/BlockHandleImpl.java [hunk 4]
```diff
@@ -54,9 +60,15 @@ public BlockCapsule produce(Miner miner, long blockTime, long timeout) {
       manager.pushBlock(blockCapsule);
       tronNetService.broadcast(blockMessage);
     } catch (Exception e) {
+      monitorMetric.getMeter(MonitorMetric.NODE_STATUS)
+          .mark();
       logger.error("Handle block {} failed.", blockCapsule.getBlockId().getString(), e);
       return null;
     }
+
+    monitorMetric.getMeter(MonitorMetric.BLOCKCHAIN_BLOCK_COUNT)
```
Comment (wubin01) [inline framework/src/main/java/org/tron/core/consensus/BlockHandleImpl.java]: BLOCKCHAIN_BLOCK_COUNT doesn't need to be defined as metrics

File: framework/src/main/java/org/tron/core/consensus/BlockHandleImpl.java [hunk 5]
```diff
@@ -54,9 +60,15 @@ public BlockCapsule produce(Miner miner, long blockTime, long timeout) {
       manager.pushBlock(blockCapsule);
       tronNetService.broadcast(blockMessage);
     } catch (Exception e) {
+      monitorMetric.getMeter(MonitorMetric.NODE_STATUS)
```
Comment (wubin01) [inline framework/src/main/java/org/tron/core/consensus/BlockHandleImpl.java]: NODE_STATUS Why is it set here ？ and doesn't need to be defined as metrics

File: framework/src/main/java/org/tron/core/consensus/BlockHandleImpl.java [hunk 6]
```diff
@@ -43,6 +47,9 @@ public Object getLock() {
   }
 
   public BlockCapsule produce(Miner miner, long blockTime, long timeout) {
+    long now = System.currentTimeMillis();
+    BlockChainInfo blockInfo=new BlockChainInfo(false);
```
[the remark this answers is not in the document]
Comment (spidemen) [reply framework/src/main/java/org/tron/core/consensus/BlockHandleImpl.java]: will remove it

File: common/src/main/java/org/tron/common/entity/NodeInfo.java (hunk 3 above)
Comment (spidemen) [reply common/src/main/java/org/tron/common/entity/NodeInfo.java]: will remove it

File: framework/src/main/java/org/tron/core/consensus/BlockHandleImpl.java (hunk 5 above)
Comment (spidemen) [reply framework/src/main/java/org/tron/core/consensus/BlockHandleImpl.java]: check node status, if there is any expection during a period, then set node status to inactive.

File: framework/src/main/java/org/tron/program/FullNode.java [hunk 7]
```diff
@@ -74,9 +74,7 @@ public static void main(String[] args) {
 
     // http api server
     FullNodeHttpApiService httpApiService = context.getBean(FullNodeHttpApiService.class);
-    if (CommonParameter.getInstance().fullNodeHttpEnable) {
-      appT.addService(httpApiService);
-    }
+    appT.addService(httpApiService);
```
[the remark this answers is not in the document]
Comment (spidemen) [reply framework/src/main/java/org/tron/program/FullNode.java]: will add it

File: framework/src/main/java/org/tron/core/metrics/MetricsService.java [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+   * set node info.
+   *
+   * @param data MetricsInfo
+   */
+  public void setNodeInfo(MetricsInfo data) {
+    MetricsInfo.NodeInfo nodeInfo = new MetricsInfo.NodeInfo();
+
+    nodeInfo.setIp(getMyIp());
+    nodeInfo.setType(1);
+    nodeInfo.setStatus(BlockChainInfo.produceBlockexpectionCount >= 1 ? 0 : 1);
```
[the remark this answers is not in the document]
Comment (spidemen) [reply framework/src/main/java/org/tron/core/metrics/MetricsService.java]: please take a look at our lastest code.

File: framework/src/main/java/org/tron/core/metrics/MetricsService.java [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+
+  /**
+   * set node info.
+   *
+   * @param data MetricsInfo
+   */
+  public void setNodeInfo(MetricsInfo data) {
+    MetricsInfo.NodeInfo nodeInfo = new MetricsInfo.NodeInfo();
+
+    nodeInfo.setIp(getMyIp());
```
[the remark this answers is not in the document]
Comment (spidemen) [reply framework/src/main/java/org/tron/core/metrics/MetricsService.java]: will take a look

File: framework/src/main/java/org/tron/core/db/Manager.java [hunk 10]
```diff
@@ -1179,6 +1193,11 @@ public TransactionInfo processTransaction(final TransactionCapsule trxCap, Block
       ownerAddressSet.add(ByteArray.toHexString(TransactionCapsule.getOwner(contract)));
     }
 
+    monitorMetric.getMeter(MonitorMetric.BLOCKCHAIN_BLOCKPROCESS_TIME)
```
Comment (wubin01) [inline framework/src/main/java/org/tron/core/db/Manager.java]: Isn't this a function of dealing with transactions

File: common/src/main/java/org/tron/core/Constant.java [hunk 11]
```diff
@@ -239,4 +239,6 @@
   public static final String RATE_LIMITER_RPC = "rate.limiter.rpc";
 
   public static final String SEED_NODE_IP_LIST = "seed.node.ip.list";
+
+  public static final String NODE_METRICS_ENABLE = "metrics.nodeMetricsEnable";
```
Comment (wubin01) [inline common/src/main/java/org/tron/core/Constant.java]: define to node. metricsEnable

File: framework/src/main/java/org/tron/core/consensus/BlockHandleImpl.java [hunk 12]
```diff
@@ -54,9 +59,12 @@ public BlockCapsule produce(Miner miner, long blockTime, long timeout) {
       manager.pushBlock(blockCapsule);
       tronNetService.broadcast(blockMessage);
     } catch (Exception e) {
+      metricsService.meterMark(MetricsService.NODE_STATUS, 1);
```
Comment (wubin01) [inline framework/src/main/java/org/tron/core/consensus/BlockHandleImpl.java]: There's something wrong with this metric， no need NODE_STATUS ， and NODE_STATUS is not a metric

Comment (wubin01) [inline framework/src/main/java/org/tron/core/consensus/BlockHandleImpl.java]: There's something wrong with this metric， no need NODE_STATUS ， and NODE_STATUS is not a metric

File: framework/src/main/java/org/tron/core/db/Manager.java [hunk 13]
```diff
@@ -1179,6 +1191,10 @@ public TransactionInfo processTransaction(final TransactionCapsule trxCap, Block
       ownerAddressSet.add(ByteArray.toHexString(TransactionCapsule.getOwner(contract)));
     }
 
+    metricsService.meterMark(MetricsService.BLOCKCHAIN_BLOCKPROCESS_TIME,
```
Comment (wubin01) [inline framework/src/main/java/org/tron/core/db/Manager.java]: This is dealing with transactions, not blocks

File: framework/src/main/java/org/tron/core/net/TronNetDelegate.java [hunk 14]
```diff
@@ -259,4 +276,35 @@ public boolean validBlock(BlockCapsule block) throws P2pException {
...
+
+  public long getFailProcessBlockNum() {
+    return failProcessBlockNum;
+  }
+
+  public String getFailProcessBlockReason() {
+    return failProcessBlockReason;
+  }
+
+  private void recordBlockLatency(BlockCapsule block, long nowTime) {
```
Comment (wubin01) [inline framework/src/main/java/org/tron/core/net/TronNetDelegate.java]: This logic should not be handled in this place

File: framework/src/main/java/org/tron/common/overlay/discover/node/NodeManager.java [hunk 15]
```diff
@@ -231,6 +236,8 @@ public void handleEvent(UdpEvent udpEvent) {
 
     NodeHandler nodeHandler = getNodeHandler(n);
     nodeHandler.getNodeStatistics().messageStatistics.addUdpInMessage(m.getType());
+    metricsService.meterMark(MetricsKey.NET_UDP_IN_TRAFFIC,
```
Comment (wubin01) [inline framework/src/main/java/org/tron/common/overlay/discover/node/NodeManager.java]: Why + 1

Comment (chengtx01) [reply framework/src/main/java/org/tron/common/overlay/discover/node/NodeManager.java]: add one byte for message type

File: framework/src/main/java/org/tron/core/metrics/blockchain/BlockChainMetricManager.java [hunk 16]
```diff
@@ -0,0 +1,230 @@
...
+    RateInfo tpsInfo = new RateInfo();
+    tpsInfo.setCount(transactionRate.getCount());
+    tpsInfo.setMeanRate(transactionRate.getMeanRate());
+    tpsInfo.setOneMinuteRate(transactionRate.getOneMinuteRate());
+    tpsInfo.setFiveMinuteRate(transactionRate.getFiveMinuteRate());
+    tpsInfo.setFifteenMinuteRate(transactionRate.getFifteenMinuteRate());
+    return tpsInfo;
+  }
+
+  private List<WitnessInfo> getNoUpgradedSR() {
```
Comment (chengtx01) [inline framework/src/main/java/org/tron/core/metrics/blockchain/BlockChainMetricManager.java]: redundant code

File: framework/src/main/java/org/tron/core/services/http/MetricsServlet.java [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+      if (metricsInfo != null) {
+        response.getWriter().println(JSON.toJSONString(metricsInfo, true));
+      } else {
+        response.getWriter().println("{}");
+      }
+    } catch (Exception e) {
+      Util.processError(e, response);
+    }
+  }
+}
```
Comment (chengtx01) [inline framework/src/main/java/org/tron/core/services/http/MetricsServlet.java]: add POST

File: framework/src/main/java/org/tron/common/application/ApplicationImpl.java [hunk 18]
```diff
@@ -31,6 +32,9 @@
   @Autowired
   private ConsensusService consensusService;
 
+  @Autowired
+  private BlockChainMetricManager blockChainMetricManager;
```
Comment (wubin01) [inline framework/src/main/java/org/tron/common/application/ApplicationImpl.java]: The setting is reversed. It should be metricsservice.init (blockchainmetricmanager. Setmetricsservice (this)))

File: framework/src/main/java/org/tron/core/services/http/MetricsServlet.java (hunk 17 above)
Comment (wubin01) [reply framework/src/main/java/org/tron/core/services/http/MetricsServlet.java]: no need post

File: protocol/src/main/protos/core/Tron.proto [hunk 19]
```diff
@@ -630,4 +630,121 @@ message NodeInfo {
       string stackTrace = 7;
     }
   }
```
Comment (wubin01) [inline protocol/src/main/protos/core/Tron.proto]: delete this proto

Comment (spidemen) [reply protocol/src/main/protos/core/Tron.proto]: ok

File: framework/src/main/java/org/tron/core/metrics/MetricsService.java [hunk 20]
```diff
@@ -0,0 +1,124 @@
...
+  public void applyBlock(BlockCapsule block) {
+    try {
+      blockChainMetricManager.applyBlock(block);
+    } catch (Exception e) {
+      logger.warn("record block failed, {}, reason: {}.",
+              block.getBlockId().toString(), e.getMessage());
+    }
+  }
+
+  public void failProcessBlock(long blockNum, String errorInfo) {
```
Comment (wubin01) [inline framework/src/main/java/org/tron/core/metrics/MetricsService.java]: Move logic to blockchainmetricmanager for processing

File: framework/src/main/java/org/tron/core/metrics/blockchain/BlockChainMetricManager.java [hunk 21]
```diff
@@ -0,0 +1,217 @@
...
+  }
+
+  private void setBlockChainInfo(BlockChainInfo blockChain) {
+    blockChain.setHeadBlockTimestamp(chainBaseManager.getHeadBlockTimeStamp());
+    blockChain.setHeadBlockHash(dbManager.getDynamicPropertiesStore()
+            .getLatestBlockHeaderHash().toString());
+
+    RateInfo blockProcessTime = getBlockProcessTime();
+    blockChain.setBlockProcessTime(blockProcessTime);
+    blockChain.setForkCount(getSuccessForkCount());
```
Comment (wubin01) [inline framework/src/main/java/org/tron/core/metrics/blockchain/BlockChainMetricManager.java]: is ForkCount， not SuccessForkCount

File: framework/src/main/java/org/tron/core/metrics/blockchain/BlockChainMetricManager.java [hunk 22]
```diff
@@ -0,0 +1,217 @@
...
+public class BlockChainMetricManager {
+
+
+  @Autowired
+  private Manager dbManager;
+
+  @Autowired
+  private ChainBaseManager chainBaseManager;
+
+  @Autowired
```
Comment (wubin01) [inline framework/src/main/java/org/tron/core/metrics/blockchain/BlockChainMetricManager.java]: Dependency is in reverse

File: framework/src/main/java/org/tron/core/metrics/blockchain/BlockChainMetricManager.java [hunk 23]
```diff
@@ -0,0 +1,217 @@
...
+    witnessInfo.put(witnessAddress, block);
+
+    //latency
+    long netTime = nowTime - block.getTimeStamp();
+    metricsService.histogramUpdate(MetricsKey.NET_BLOCK_LATENCY, netTime);
+    metricsService.histogramUpdate(MetricsKey.NET_BLOCK_LATENCY_WITNESS + witnessAddress, netTime);
+    if (netTime >= 1000) {
+      metricsService.counterInc(MetricsKey.NET_BLOCK_LATENCY + ".1S", 1L);
+      metricsService.counterInc(MetricsKey.NET_BLOCK_LATENCY_WITNESS + witnessAddress + ".1S", 1L);
+      if (netTime >= 2000) {
```
Comment (wubin01) [inline framework/src/main/java/org/tron/core/metrics/blockchain/BlockChainMetricManager.java]: wrong logic ，should be  'else if'

File: framework/src/main/java/org/tron/core/metrics/blockchain/BlockChainMetricManager.java [hunk 24]
```diff
@@ -0,0 +1,217 @@
...
+        BlockCapsule block = witnessInfo.get(address);
+        WitnessInfo witness = new WitnessInfo(address,
+                block.getInstance().getBlockHeader().getRawData().getVersion());
+        witnessInfos.add(witness);
+      }
+    }
+    return witnessInfos;
+  }
+
+  private RateInfo getBlockProcessTime() {
```
Comment (wubin01) [inline framework/src/main/java/org/tron/core/metrics/blockchain/BlockChainMetricManager.java]: Common method required

File: framework/src/main/java/org/tron/core/metrics/blockchain/BlockChainMetricManager.java [hunk 25]
```diff
@@ -0,0 +1,217 @@
...
+    tpsInfo.setMeanRate(transactionRate.getMeanRate());
+    tpsInfo.setOneMinuteRate(transactionRate.getOneMinuteRate());
+    tpsInfo.setFiveMinuteRate(transactionRate.getFiveMinuteRate());
+    tpsInfo.setFifteenMinuteRate(transactionRate.getFifteenMinuteRate());
+    return tpsInfo;
+  }
+
+
+  // gap: 1 minute, 5 minute, 15 minute, 0: avg for total block and time
+  private double getAvgBlockProcessTimeByGap(int gap) {
```
Comment (wubin01) [inline framework/src/main/java/org/tron/core/metrics/blockchain/BlockChainMetricManager.java]: no need conversion, This function is useless. delete it

File: framework/src/main/java/org/tron/core/metrics/blockchain/BlockChainMetricManager.java [hunk 26]
```diff
@@ -0,0 +1,217 @@
...
+        int gapFifteenTimeCount =
+            Math.round(Math.round(meterBlockTxCount.getFifteenMinuteRate() * gap * 60));
+        return gapFifteenTimeBlock / (double) gapFifteenTimeCount;
+
+      default:
+        return -1;
+    }
+  }
+
+  private int getSuccessForkCount() {
```
Comment (wubin01) [inline framework/src/main/java/org/tron/core/metrics/blockchain/BlockChainMetricManager.java]: Use public methods, no need to encapsulate

File: framework/src/main/java/org/tron/core/metrics/blockchain/BlockChainMetricManager.java [hunk 27]
```diff
@@ -0,0 +1,217 @@
...
+    return (int) metricsService.getMeter(MetricsKey.BLOCKCHAIN__FORK_COUNT).getCount();
+  }
+
+  private int getFailForkCount() {
```
Comment (wubin01) [inline framework/src/main/java/org/tron/core/metrics/blockchain/BlockChainMetricManager.java]: Use public methods, no need to encapsulate

File: framework/src/main/java/org/tron/core/metrics/blockchain/StartTimeRecorder.java [hunk 28]
```diff
@@ -0,0 +1,23 @@
+package org.tron.core.metrics.blockchain;
+
+public class StartTimeRecorder {
```
Comment (wubin01) [inline framework/src/main/java/org/tron/core/metrics/blockchain/StartTimeRecorder.java]: delete this class

File: framework/src/main/java/org/tron/core/metrics/blockchain/BlockChainMetricManager.java (hunk 21 above)
Comment (spidemen) [reply framework/src/main/java/org/tron/core/metrics/blockchain/BlockChainMetricManager.java]: ok

File: framework/src/main/java/org/tron/core/metrics/blockchain/BlockChainMetricManager.java (hunk 24 above)
Comment (spidemen) [reply framework/src/main/java/org/tron/core/metrics/blockchain/BlockChainMetricManager.java]: ok

File: framework/src/main/java/org/tron/core/metrics/blockchain/BlockChainMetricManager.java (hunk 25 above)
Comment (spidemen) [reply framework/src/main/java/org/tron/core/metrics/blockchain/BlockChainMetricManager.java]: no. This function is needed in order to keep more precious avg block process time

File: framework/src/main/java/org/tron/core/metrics/blockchain/BlockChainMetricManager.java (hunk 27 above)
Comment (spidemen) [reply framework/src/main/java/org/tron/core/metrics/blockchain/BlockChainMetricManager.java]: ok

File: framework/src/main/java/org/tron/core/metrics/blockchain/BlockChainMetricManager.java (hunk 26 above)
Comment (spidemen) [reply framework/src/main/java/org/tron/core/metrics/blockchain/BlockChainMetricManager.java]: ok

**Pull request closed without merging**